aws_crt_common_ws/aws-crt-common/src/lib.rs

```diff
@@ -0,0 +1,596 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+use aws_crt_common_sys::{
+    aws_common_library_clean_up, aws_common_library_init, aws_default_allocator,
+    aws_error_debug_str, aws_error_str, aws_last_error, aws_mem_tracer_bytes,
+    aws_mem_tracer_destroy, aws_mem_tracer_new, aws_register_error_info, aws_string_bytes,
+    aws_string_destroy, aws_string_new_from_array, aws_unregister_error_info, CAllocator, CString,
+    TraceLevel,
+};
+use futures::task::Waker;
+use std::ffi::CStr;
+use std::future::Future;
+use std::os::raw::{c_char, c_void};
+use std::pin::Pin;
+use std::ptr::null_mut;
+use std::str::Utf8Error;
+use std::sync::{Arc, Mutex, MutexGuard};
+use std::task::{Context, Poll};
+
+/// Trait for deep copying (cloning) an object. The operation can fail so it returns
+/// a result with an error code upon failure.
+pub trait TryClone<T> {
+    fn try_clone(&self) -> Result<T, i32>;
+}
+
+/// Trait that gives a c-style non-const pointer to an underlying resource.
+/// Ideally you use this when calling an ffi function that needs a pointer
+/// to your wrapped Rust object.
+pub trait MutableNativeResource<T> {
+    fn get_native_resource_mut(&mut self) -> *mut T;
+}
+
+/// Trait that gives a c-style const pointer to an underlying resource.
+/// Ideally you use this when calling an ffi function that needs a pointer
+/// to your wrapped Rust object.
+pub trait NativeResource<T> {
+    fn get_native_resource(&self) -> *const T;
+}
+
+pub struct FutureSharedState<T>
+where
+    T: Clone,
+{
+    data: Option<T>,
+    waker: Option<Waker>,
+}
+
+/// Simple future implementation for driving async futures
+/// from C-style callbacks. For this scenario, a pointer to FutureSharedState is used as your user_data pointer.
+///
+/// Upon creation call, CallbackFuture::to_callback_ptr() to get the pointer to pass to C functions.
+/// When your callback completes, fill-in an instance of type T, and call FutureSharedState::resolve()
+/// with the user_data pointer from your c callback.
+///
+/// Assuming you do these things, you've just converted C-style async code into co-routine ready
+/// Rust futures.
+///
+#[derive(Clone)]
+pub struct CallbackFuture<T>
```

Comment:
This is pretty much Tokio's [`oneshot` channel](https://docs.rs/tokio/1.6.0/tokio/sync/oneshot/index.html). The only difference is you have a conversion to raw pointers. We could add to/from raw pointers to the Tokio channel as this is a pretty common pattern when bridging to C. It also is possible to only depend on Tokio's sync primitives w/o pulling in the rest if that is a path you want to explore.